File: app/src/main/java/com/doapps/habits/fragments/CreateFragment.java
package com.doapps.habits.fragments;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import com.doapps.habits.BuildConfig;
import com.doapps.habits.R;
import com.doapps.habits.activity.MainActivity;
import com.doapps.habits.dao.HabitDao;
import com.doapps.habits.helper.HabitListManager;
import com.doapps.habits.helper.HabitNotifier;
import com.doapps.habits.models.Habit;
import java.util.Arrays;
import java.util.Calendar;

public class CreateFragment extends Fragment implements AdapterView.OnItemSelectedListener {

  private TextInputEditText editTime;
  private TextInputEditText editQuestion;
  private TextInputEditText editTitle;
  private Spinner sFrequency;
  private TextView tvFreqNum;
  private TextView tvFreqDen;
  private ViewGroup llCustomFrequency;
  private int lastSpinnerSelection;
  private HabitDao habitsDatabase;

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    View result = inflater.inflate(R.layout.fragment_create, container, false);
    editTitle = result.findViewById(R.id.edit_title);
    editQuestion = result.findViewById(R.id.edit_question);
    editTitle.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        editTitle.setError("Can't be empty");
      }

      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        if (charSequence.length() != 0) {
          char start = Character.toLowerCase(charSequence.charAt(0));
          editQuestion.setText(String.format("Did you %s%s today?", start,
              charSequence.subSequence(1, charSequence.length())));
        }
      }

      @Override
      public void afterTextChanged(Editable editable) {
        if (editable.length() == 0) {
          editTitle.setError("Can't be empty");
        }
      }
    });
    editTime = result.findViewById(R.id.edit_time);
    sFrequency = result.findViewById(R.id.sFrequency);
    tvFreqNum = result.findViewById(R.id.input_freq_num);
    tvFreqDen = result.findViewById(R.id.input_freq_den);
    llCustomFrequency = result.findViewById(R.id.llCustomFrequency);

    sFrequency.setOnItemSelectedListener(this);
    result.findViewById(R.id.btnCreate).setOnClickListener(view -> onHabitCreate());
    ((MainActivity) getActivity()).getToolbar().setTitle(R.string.habit_create);

    return result;
  }

  private void onHabitCreate() {
    if (editTitle.length() != 0) {
      int[] frequency;
      if (lastSpinnerSelection == 0) {
        frequency = new int[]{1, 1};
      } else if (lastSpinnerSelection == 1) {
        frequency = new int[]{7, 1};
      } else if (lastSpinnerSelection == 2) {
        frequency = new int[]{7, 2};
      } else if (lastSpinnerSelection == 3) {
        frequency = new int[]{7, 5};
      } else {
        frequency = new int[2];
        frequency[0] = Integer.valueOf(tvFreqDen.getText().toString());
        frequency[1] = Integer.valueOf(tvFreqNum.getText().toString());
      }

      if (BuildConfig.DEBUG) {
        Log.i("onHabitCreate", "freq = " + Arrays.toString(frequency));
      }

      int time = editTime.getText().toString().isEmpty() ?
          0 : Integer.valueOf(editTime.getText().toString());
      // Checks what data was entered and adds habit to habitsDatabase
      habitsDatabase = HabitListManager.getInstance(getContext()).getDatabase().habitDao();
      Calendar upd = Calendar.getInstance();
      Habit habit = new Habit(
          editTitle.getText().toString(),
          editQuestion.getText().toString(),
          false,
          upd.get(Calendar.DATE),
          upd.get(Calendar.MONTH),
          upd.get(Calendar.YEAR),
          time,
          System.currentTimeMillis(),
          0, frequency);
      new InsertHabitTask().execute(habit);
    } else {
      Toast.makeText(getContext().getApplicationContext(), "Please enter habit's title",
          Toast.LENGTH_SHORT).show();
    }
  }

  @Override
  public void onItemSelected(AdapterView<?> adapterView, View view, int i,
      long l) {
    lastSpinnerSelection = i;
    if (i == 4) {
      sFrequency.setVisibility(View.GONE);
      llCustomFrequency.setVisibility(View.VISIBLE);
    }
  }

  @Override
  public void onNothingSelected(AdapterView<?> adapterView) {
    // ignored
  }

  @Override
  public void onPause() {
    ((MainActivity) getActivity()).closeImm();
    super.onPause();
  }

  @SuppressLint("StaticFieldLeak")
  private class InsertHabitTask extends AsyncTask<Habit, Void, Long> {

    @Override
    protected Long doInBackground(Habit... habit) {
      Log.i("DatabaseContains", "New program added");
      return habitsDatabase.insert(habit[0]);
    }

    @Override
    protected void onPostExecute(Long id) {
      HabitNotifier habitNotifier = new HabitNotifier(getContext(),
          editQuestion.getText().toString(), id);
      habitNotifier.initiate();

      ((MainActivity) getActivity()).getToolbar().setTitle("List");
      getFragmentManager().beginTransaction().replace(R.id.content_frame, new ListFragment())
          .commit();
      super.onPostExecute(id);
    }
  }
}

File: app/src/main/java/com/doapps/habits/fragments/CreateFragment.kt
package com.doapps.habits.fragments

import android.annotation.SuppressLint
import android.os.AsyncTask
import android.os.Bundle
import android.support.design.widget.TextInputEditText
import android.support.v4.app.Fragment
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import com.doapps.habits.BuildConfig
import com.doapps.habits.R
import com.doapps.habits.activity.MainActivity
import com.doapps.habits.database.HabitsDatabase
import com.doapps.habits.helper.HabitListManager
import com.doapps.habits.models.Habit
import com.doapps.habits.receivers.NotificationReceiver
import java.util.*

class CreateFragment : Fragment(), AdapterView.OnItemSelectedListener {

  private lateinit var editTime: TextInputEditText
  private lateinit var editQuestion: TextInputEditText
  private lateinit var editTitle: TextInputEditText
  private lateinit var sFrequency: Spinner
  private lateinit var tvFreqNum: TextView
  private lateinit var tvFreqDen: TextView
  private lateinit var llCustomFrequency: ViewGroup
  private var lastSpinnerSelection: Int = 0
  private var habitsDatabase: HabitsDatabase? = null

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    // Inflate the layout for this fragment
    val result = inflater.inflate(R.layout.fragment_create, container, false)
    editTitle = result.findViewById(R.id.edit_title)
    editQuestion = result.findViewById(R.id.edit_question)
    editTitle.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
        editTitle.error = "Can't be empty"
      }

      override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
        if (charSequence.isNotEmpty()) {
          val start = Character.toLowerCase(charSequence[0])
          editQuestion.setText(String.format("Did you %s%s today?", start,
              charSequence.subSequence(1, charSequence.length)))
        }
      }

      override fun afterTextChanged(editable: Editable) {
        if (editable.isEmpty()) {
          editTitle.error = "Can't be empty"
        }
      }
    })
    editTime = result.findViewById(R.id.edit_time)
    sFrequency = result.findViewById(R.id.sFrequency)
    tvFreqNum = result.findViewById(R.id.input_freq_num)
    tvFreqDen = result.findViewById(R.id.input_freq_den)
    llCustomFrequency = result.findViewById(R.id.llCustomFrequency)

    sFrequency.onItemSelectedListener = this
    result.findViewById<View>(R.id.btnCreate).setOnClickListener { onHabitCreate() }
    (activity as MainActivity).toolbar.setTitle(R.string.habit_create)

    return result
  }

  private fun onHabitCreate() {
    if (editTitle.length() != 0) {
      val frequency = when (lastSpinnerSelection) {
        0 -> intArrayOf(1, 1)
        1 -> intArrayOf(1, 7)
        2 -> intArrayOf(2, 7)
        3 -> intArrayOf(5, 7)
        else -> intArrayOf(Integer.valueOf(tvFreqNum.text.toString())!!, Integer.valueOf(tvFreqDen.text.toString())!!)
      }

      if (BuildConfig.DEBUG) {
        Log.i("onHabitCreate", "freq = " + Arrays.toString(frequency))
      }

      val time = when {
        editTime.text.toString().isNotEmpty() -> Integer.valueOf(editTime.text.toString())
        else -> 0
      }
      // Checks what data was entered and adds habit to habitsDatabase
      habitsDatabase = HabitListManager.getInstance(context).habitsDatabase
      val upd = Calendar.getInstance()
      val habit = Habit(
          editTitle.text.toString(),
          editQuestion.text.toString(),
          false,
          upd.get(Calendar.DATE),
          upd.get(Calendar.MONTH),
          upd.get(Calendar.YEAR),
          time,
          System.currentTimeMillis(),
          0, *frequency)
      InsertHabitTask().execute(habit)
    } else {
      Toast.makeText(context!!.applicationContext, "Please enter habit's title", Toast.LENGTH_SHORT).show()
    }
  }

  override fun onItemSelected(adapterView: AdapterView<*>, view: View?, i: Int, l: Long) {
    if (view != null) {
      lastSpinnerSelection = i
      if (i == 4) {
        sFrequency.visibility = View.GONE
        llCustomFrequency.visibility = View.VISIBLE
      }
    }
//    else {
//      sFrequency.visibility = View.VISIBLE
//      llCustomFrequency.visibility = View.GONE
//    }
  }

  override fun onNothingSelected(adapterView: AdapterView<*>) {
    // ignored
  }

  override fun onPause() {
    (activity as MainActivity).closeImm()
    super.onPause()
  }

  @SuppressLint("StaticFieldLeak")
  private inner class InsertHabitTask : AsyncTask<Habit, Unit, Long>() {

    override fun doInBackground(vararg habit: Habit): Long? {
      if (BuildConfig.DEBUG) {
        Log.i("InsertHabitTask", "New program added")
      }
      return habitsDatabase!!.habitDao().insert(habit[0])
    }

    override fun onPostExecute(id: Long?) {
      val notificationService = NotificationReceiver()
      notificationService.setAlarm(context!!, id!!, editQuestion.text.toString())
      (activity as MainActivity).toolbar.title = getString(R.string.list)
      (activity as MainActivity).setChecked(2)
      (activity as MainActivity).mLastFragment = R.id.nav_list

      fragmentManager!!.beginTransaction().replace(R.id.content_frame, ListFragment())
          .commit()
      super.onPostExecute(id)
    }
  }
}
